crates/ruma-signatures/src/error.rs

```diff
@@ -0,0 +1,276 @@
+    pub fn public_key_not_found<T: Into<String>>(target: T) -> Error {
+        Self::PublicKeyNotFound(target.into()).into()
+    }
+}
+
+/// Errors relating to parsing of all sorts.
+#[derive(Debug, Error)]
+#[non_exhaustive]
+pub enum ParseError {
+    /// For user ID parsing errors.
+    #[error("Could not parse User ID: {0}")]
+    UserId(#[source] ruma_identifiers::Error),
+    /// For event ID parsing errors.
+    #[error("Could not parse Event ID: {0}")]
+    EventId(#[source] ruma_identifiers::Error),
+
+    /// For when an event ID, coupled with a specific room version, doesn't have a server name
+    /// embedded.
+    #[error("Event Id {0:?} should have a server name for the given room version {1:?}")]
+    ServerNameFromEventIdByRoomVersion(EventId, RoomVersionId),
+
+    /// For when the extracted/"parsed" public key from a PKCS#8 v2 document doesn't match the
+    /// public key derived from it's private key.
+    #[error("PKCS#8 Document public key does not match public key derived from private key; derived: {0:X?} (len {}), parsed: {1:X?} (len {})", .derived_key.len(), .parsed_key.len())]
+    DerivedPublicKeyDoesNotMatchParsedKey {
+        /// The parsed key.
+        parsed_key: Vec<u8>,
+        /// The derived key.
+        derived_key: Vec<u8>,
+    },
+
+    /// For when the ASN.1 Object Identifier on a PKCS#8 document doesn't match the expected one.
+    ///
+    /// e.g. the document describes a RSA key, while an ed25519 key was expected.
+    #[error("Algorithm OID does not match ed25519, expected {expected}, found {found}")]
+    Oid {
+        /// The expected OID.
+        expected: pkcs8::ObjectIdentifier,
+        /// The OID that was found instead.
+        found: pkcs8::ObjectIdentifier,
+    },
+
+    /// For when [`ed25519_dalek`] cannot parse a secret/private key.
+    #[error("Could not parse secret key: {0}")]
+    SecretKey(#[source] ed25519_dalek::SignatureError),
+
+    /// For when [`ed25519_dalek`] cannot parse a public key.
+    #[error("Could not parse public key: {0}")]
+    PublicKey(#[source] ed25519_dalek::SignatureError),
+
+    /// For when [`ed25519_dalek`] cannot parse a signature.
+    #[error("Could not parse signature: {0}")]
+    Signature(#[source] ed25519_dalek::SignatureError),
+
+    /// For when parsing base64 gives an error.
+    #[error("Could not parse {of_type} base64 string {string:?}: {source}")]
+    Base64 {
+        /// The "type"/name of the base64 string
+        of_type: String,
+        /// The string itself.
+        string: String,
+        /// The originating error.
+        #[source]
+        source: base64::DecodeError,
+    },
+}
+
+impl ParseError {
+    /// Convenience method for [`ParseError::ServerNameFromEventIdByRoomVersion`].
+    pub fn from_event_id_by_room_version(
+        event_id: &EventId,
+        room_version: &RoomVersionId,
+    ) -> Error {
+        Self::ServerNameFromEventIdByRoomVersion(
+            // FIX: this can be made better
```

Comment:
How you wanted to made this better?

Reply:
It's two `.clone()`s, I think that's a bit of a code smell, thus the comment

Reply:
Well, or either you receive them owned (so, the clone happens on the caller site) or you clone it here. I don't think that there's any issue with those clones and I would probably remove this comment.